reqwest-tracing/src/reqwest_otel_span_macro.rs

```diff
@@ -0,0 +1,170 @@
+        {
+            let method = $request.method();
+            let scheme = $request.url().scheme();
+            let host = $request.url().host_str().unwrap_or("");
+            let host_port = $request.url().port().unwrap_or(0) as i64;
+            let path = $request.url().path();
+            let otel_name = format!("{} {}", method, path);
+
+            macro_rules! request_span {
+                ($lvl:expr) => {
+                    $crate::reqwest_otel_span_macro::private::span!(
+                        $lvl,
+                        "HTTP request",
+                        http.method = %method,
+                        http.scheme = %scheme,
+                        http.host = %host,
+                        net.host.port = %host_port,
+                        otel.kind = "client",
+                        otel.name = %otel_name,
+                        otel.status_code = tracing::field::Empty,
+                        http.user_agent = tracing::field::Empty,
+                        http.status_code = tracing::field::Empty,
+                        error.message = tracing::field::Empty,
+                        error.cause_chain = tracing::field::Empty,
+                        $($field)*
+                    )
+                }
+            }
+
+            let span = match $level {
+                $crate::reqwest_otel_span_macro::private::Level::TRACE => {
+                    request_span!($crate::reqwest_otel_span_macro::private::Level::TRACE)
+                },
+                $crate::reqwest_otel_span_macro::private::Level::DEBUG => {
+                    request_span!($crate::reqwest_otel_span_macro::private::Level::DEBUG)
+                },
+                $crate::reqwest_otel_span_macro::private::Level::INFO => {
+                    request_span!($crate::reqwest_otel_span_macro::private::Level::INFO)
+                },
+                $crate::reqwest_otel_span_macro::private::Level::WARN => {
+                    request_span!($crate::reqwest_otel_span_macro::private::Level::WARN)
+                },
+                $crate::reqwest_otel_span_macro::private::Level::ERROR => {
+                    request_span!($crate::reqwest_otel_span_macro::private::Level::ERROR)
+                },
+            };
+            span
+        }
+    }
+}
+
+#[macro_export]
+/// provides a default [`on_request_success`] implementation for a [`reqwest_otel_span!`] span.
+///
+/// [`on_request_success`]: crate::RequestOtelSpanBackend::on_request_success
+macro_rules! impl_on_request_success {
+    () => {
+        fn on_request_success(span: &Span, response: &Response, _: &mut Extensions) {
+            crate::default_on_request_success(span, response)
+        }
+    };
+}
+
+#[macro_export]
+/// provides a default [`on_request_failure`] implementation for a [`reqwest_otel_span!`] span.
+///
+/// [`on_request_failure`]: crate::RequestOtelSpanBackend::on_request_failure
+macro_rules! impl_on_request_failure {
```

Comment:
If the fields are not there, the `record` call will simply desugar to a no-op. I don't think that's an issue.

Reply:
Suppose an implementor defines a field with a default value and a name the default implementation writes to. In that case, it could be an issue because the default implementation will overwrite it.

Reply:
I see. I don't see the need for a macro though - we can simply provide default function handlers.